qubes/devices.py

```diff
@@ -115,87 +142,113 @@ class DeviceCollection(object):
+            device_assignment.persistent = True
+            self._set.discard(device_assignment)
+
         self._vm.fire_event('device-detach:' + self._class, device=device)
 
-    def attached(self, persistent=None):
-        '''List devices which are (or may be) attached to this vm
+    def attached(self):
+        '''List devices which are (or may be) attached to this vm '''
+        return self._vm.fire_event('device-list-attached:' + self._class) or []
+
+    def persistent(self):
+        ''' Devices persistently attached and safe to access before libvirt
+            bootstrap.
+        '''
+        return [self._device(a) for a in self._set]
+
+    def assignments(self, persistent=None):
+        '''List assignments for devices which are (or may be) attached to the
+           vm.
 
         Devices may be attached persistently (so they are included in
         :file:`qubes.xml`) or not. Device can also be in :file:`qubes.xml`,
         but be temporarily detached.
 
-        :param bool persistent: only include devices which are (or are not) \
-        attached persistently - None means both
+        :param bool persistent: only include devices which are or are not
+        attached persistently.
         '''
-        seen = self._set.copy()
-
-        # ask for really attached devices only when requested not only
-        # persistent ones
-        if persistent is not True:
-            attached = self._vm.fire_event(
-                'device-list-attached:' + self._class,
-                persistent=persistent)
-            for device in attached:
-                device_persistent = device in self._set
-                if persistent is not None and device_persistent != persistent:
-                    continue
-                assert device.frontend_domain == self._vm, \
-                    '{!r} != {!r}'.format(device.frontend_domain, self._vm)
-
-                yield device
-
-                try:
-                    seen.remove(device)
-                except KeyError:
-                    pass
-
-        if persistent is False:
-            return
-
-        for device in seen:
-            # get fresh object - may contain updated information
-            device = device.backend_domain.devices[self._class][device.ident]
-            yield device
+
+        devices = self._vm.fire_event('device-list-attached:' + self._class,
+            persistent=persistent)
+        result = []
+        for dev in devices:
+            if dev in self._set and persistent is False:
+                continue
+            elif dev in self._set:
+                result.append(self._set.get(dev))
+            elif dev not in self._set and persistent is True:
+                continue
+            else:
+                result.append(
+                    DeviceAssignment(backend_domain=dev.backend_domain,
+                                     ident=dev.ident, options=dev.options,
+                                     frontend_domain=self._vm))
```

Comment:
Using `dev.options` here looks strange - options are property of assignment, not device itself. Better would be to expect event `device-list-attached` to return a tuples (device_info, options).